lib/core/async/interface.dart

```diff
@@ -0,0 +1,174 @@
+  }
+
+  /// Sends a response to this message with given [args].
+  void respond(SendPort port, Object? args) =>
+      InterfaceMessage(name, args, true, _id).send(port, isResponse: true);
+
+  /// Checks if the received data is a response to this [InterfaceMessage].
+  InterfaceMessage? _match(dynamic message) {
+    final response = InterfaceMessage.from(message);
+
+    if (response != null && response._isResponse && response._id == _id) {
+      return response;
+    }
+
+    return null;
+  }
+}
+
+typedef _Expectation = Tuple2<InterfaceMessage, void Function(InterfaceMessage)>;
+
+/// Interface for two-way inter-isolate communication.
+abstract class Interface {
+  /// The isolate this [Interface] is running in with
+  /// ports to the partner isolate.
+  SpawnedIsolate isolate;
+
+  /// Completes when the send port is connected.
+  ///
+  /// [ready] is set when a sent port is received on the receive port.
+  /// When setting the send port manually, ensure to `complete` [ready].
+  final Completer<void> ready = Completer();
+
+  /// A list of expected [InterfaceMessage]s and a callback when the message is received.
+  final List<_Expectation> _expectations = [];
+
+  Interface(this.isolate) {
+    // Sole listener to the receive port.
+    //? If another listener is attached a runtime error will occure.
+    isolate.receivePort.listen((dynamic message) {
+      //* Store `SendPort`s when received.
+      if (message is SendPort) {
+        isolate.sendPort = message;
+        ready.complete();
+        return;
+      }
+
+      //* Check if the received message is a response to an expected message.
+      for (final expectation in _expectations) {
+        final InterfaceMessage? response = expectation.item1._match(message);
+
+        if (response != null) {
+          expectation.item2.call(response);
+          _expectations.removeWhere((exepc) => exepc.item1._id == message._id);
+          return;
+        }
+      }
+
+      //* If the message is not a response, it is a request.
+      final request = InterfaceMessage.from(message);
+
+      if (request != null) {
+        onMessage(request);
+      }
+    });
+  }
+
+  /// Sends the [message] and returns the response if [expectResponse] is `true`.
+  Future<InterfaceMessage?> call(
+    InterfaceMessage message, {
+    bool expectResponse = false,
+    Duration? timeout,
+  }) async {
+    await ready.future;
+
+    message.send(isolate.sendPort);
+
+    if (!expectResponse) {
+      return null;
+    }
+
+    final completer = Completer<InterfaceMessage>();
+
+    if (timeout != null) {
+      Future.delayed(timeout).then((dynamic val) {
+        if (!completer.isCompleted) {
+          _expectations.removeWhere((exepc) => exepc.item1._id == message._id);
+          completer.completeError(
+            StateError(
+              'Request "${message.name}" {id: ${message._id}} timed out after ${timeout.toString()}',
+            ),
+          );
+        }
+      });
```

Comment:
```suggestion
      },);
```
And then save for formatting (I would prefer that, your decision)